core/rest-client/src/http_client.rs

```diff
@@ -0,0 +1,425 @@
+				add_to_headers(
+					&mut request_headers,
+					CONTENT_TYPE,
+					HeaderValue::from_str("application/json").unwrap(),
+				);
+
+				trace!("set request body: {}", body);
+				request.body(body_encoded.as_slice());
+			}
+		} else {
+			debug!("no body to send");
+		}
+
+		if let Some(ref auth) = self.authorization {
+			add_to_headers(
+				&mut request_headers,
+				AUTHORIZATION,
+				HeaderValue::from_str(auth).map_err(|_| Error::RequestError)?,
+			);
+		}
+
+		// add pre-set headers
+		for (key, value) in self.headers.iter() {
+			request_headers.insert(key, &value.clone());
+		}
+
+		// add user agent header
+		let pkg_version = env!("CARGO_PKG_VERSION");
+		add_to_headers(
+			&mut request_headers,
+			USER_AGENT,
+			HeaderValue::from_str(format!("integritee/{}", pkg_version).as_str())
+				.map_err(|_| Error::RequestError)?,
+		);
+
+		request.headers(HashMap::from(request_headers));
+
+		request
+			.timeout(self.timeout)
+			.connect_timeout(self.timeout)
+			.read_timeout(self.timeout)
+			.write_timeout(self.timeout);
+
+		trace!("{:?}", request);
+
+		let mut writer = Vec::new();
+
+		let response = request.send(&mut writer).map_err(Error::HttpReqError)?;
+
+		Ok((response, writer))
+	}
+}
+
+fn join_url(base_url: Url, path: &str, params: Option<&Query>) -> Result<Url, Error> {
+	let mut url = base_url.join(path).map_err(|_| Error::UrlError)?;
+
+	if let Some(params) = params {
+		for &(key, item) in params.iter() {
+			url.query_pairs_mut().append_pair(key, item);
+		}
+	}
+
+	Ok(url)
+}
+
+fn add_to_headers(headers: &mut Headers, key: HeaderName, value: HeaderValue) {
+	let header_value_str = value.to_str();
+
+	match header_value_str {
+		Ok(v) => {
+			headers.insert(key.as_str(), v);
+		},
+		Err(e) => {
+			error!("Failed to add header to request: {:?}", e);
+		},
+	}
+}
+
+#[cfg(test)]
+mod tests {
+
+	use super::*;
+	use http::header::CONNECTION;
+	use serde::{Deserialize, Serialize};
+	use std::vec::Vec;
+
+	#[test]
+	fn join_url_adds_query_parameters() {
+		let base_url = Url::parse("https://example.com").unwrap();
+		let path = "api/v2/example_list";
+		let query = [("filter", "all"), ("order", ("desc"))];
+
+		let complete_url = join_url(base_url, path, Some(&query)).unwrap();
+
+		assert_eq!(
+			complete_url.as_str(),
+			"https://example.com/api/v2/example_list?filter=all&order=desc"
+		);
+	}
+
+	#[test]
+	fn join_url_has_no_query_parameters() {
+		let base_url = Url::parse("https://example.com").unwrap();
+		let path = "api/v2/endpoint";
+		let complete_url = join_url(base_url, path, None).unwrap();
+		assert_eq!(complete_url.as_str(), "https://example.com/api/v2/endpoint");
+	}
+
+	#[test]
+	fn join_url_with_too_many_slashes() {
+		let base_url = Url::parse("https://api.mydomain.com").unwrap();
+		let path = "/api/v1/post";
+		let complete_url = join_url(base_url, path, None).unwrap();
+		assert_eq!(complete_url.as_str(), "https://api.mydomain.com/api/v1/post");
+	}
+
+	// #[test]
```

Comment:
What's up with this?

Reply:
This was a test I setup using `free.beeceptor.com`. But that instance of the REST API only runs, as long as I have it active. So it might better to just remove the test completely at this point.